lib/truffle/java/java.rb

```diff
@@ -0,0 +1,207 @@
+
+module Java
+  class JavaObject
+
+    def initialize(a_value)
+      if a_value != nil && a_value.class != Truffle::Interop::Foreign
+        raise ArgumentError, "Can only wrap native java objects"
+      end
+      @value = a_value
+    end
+    
+    def to_s
+      return @value.java_send("toString") if @value != nil
+      ""
+    end
+
+    def wrap obj
+      return JavaObject.new(obj.value) if obj.is_kind_of?(JAvaObject)
+    end
+
+    def hash
+      @value.hashCode() if @value != nil
+      0
+    end
+
+    def eql?(another)
+      return @value == @value
+    end
+
+    alias_method :eql?, :==
+
+    def equal?(another)
+      return false if @value == nil
+      return @value.equals(another.value) if another.is_kind_of?(JavaObject)
+      return false
+    end
+
+    def java_type
+      @value.getClass().getName()
+    end
+
+    def java_class
+      JavaClass.get(@value.getClass())
+    end
+
+    def length
+      raise TypeError, "Not a java array"
+    end
+
+    def java_proxy?
+      true
+    end
+
+    def synchronized(&block)
+      return @value.__synchronize__(block) if @value != nil
+      NULL_LOCK.synchronize block
+    end
+
+    def marshal_dump
+    end
+
+    def marshal_load
+    end
+    
+    private
+
+    NULL_LOCK = Mutex.new
+    
+    attr_reader :value
+    
+  end
+
+  class JavaArray < JavaObject
+    def legnth
+      java.lang.reflect.Array.getLength(@value)
+    end
+
+    def hash
+      17 * super.hash
```

Comment:
To match JRuby, but given the amount of cruft in their codebase I'm going to remove it until we know we need it.